agent/structs/config_entry.go

```diff
@@ -713,15 +756,48 @@ func (cfg UpstreamConfig) MergeInto(dst map[string]interface{}) {
 	}
 }
 
-func (cfg *UpstreamConfig) Normalize() {
+func (cfg *UpstreamConfig) NormalizeWithoutName() error {
+	return cfg.normalize(false, nil)
+}
+func (cfg *UpstreamConfig) NormalizeWithName(entMeta *EnterpriseMeta) error {
+	return cfg.normalize(true, entMeta)
+}
+func (cfg *UpstreamConfig) normalize(named bool, entMeta *EnterpriseMeta) error {
+	if named {
+		// If the upstream namespace is omitted it inherits that of the enclosing
+		// config entry.
+		cfg.EnterpriseMeta.MergeNoWildcard(entMeta)
+		cfg.EnterpriseMeta.Normalize()
+	}
+
 	cfg.Protocol = strings.ToLower(cfg.Protocol)
 
 	if cfg.ConnectTimeoutMs < 0 {
 		cfg.ConnectTimeoutMs = 0
 	}
+	return nil
 }
 
-func (cfg UpstreamConfig) Validate() error {
+func (cfg UpstreamConfig) ValidateWithoutName() error {
```

Comment:
Can you add validation/normalization tests to `structs` or a TODO for them?

Reply:
added some tests for the untested part